### PR TITLE
chore: simplify addvote

### PR DESCRIPTION
**Pull request by faddat**
This PR simplifies the AddVote function, working towards the goal of enabling the maintidx linter

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [ ] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

File: internal/consensus/state.go [hunk 1]
```diff
@@ -2348,124 +2400,122 @@ func (cs *State) addVote(vote *types.Vote, peerID p2p.ID) (added bool, err error
...
+func (cs *State) updateMetricsAndEvents(vote *types.Vote) {
 	if vote.Round == cs.Round {
 		vals := cs.state.Validators
 		_, val := vals.GetByIndex(vote.ValidatorIndex)
 		cs.metrics.MarkVoteReceived(vote.Type, val.VotingPower, vals.TotalVotingPower())
 	}
 
 	if err := cs.eventBus.PublishEventVote(types.EventDataVote{Vote: vote}); err != nil {
-		return added, err
+		cs.Logger.Error("failed to publish event vote", "err", err)
```
Comment (melekes) [inline internal/consensus/state.go:2425]: Are we ignoring this error now?

**Review by melekes — APPROVED** [review]
Thanks @faddat ❤️

File: internal/consensus/state.go [hunk 2]
```diff
@@ -2254,7 +2254,54 @@ func (cs *State) tryAddVote(vote *types.Vote, peerID p2p.ID) (bool, error) {
...
+		cs.handlePrecommit(vote)
+	default:
+		panic(fmt.Sprintf("unexpected vote type %v", vote.Type))
+	}
+
+	return true, nil
+}
+
+// logVoteDetails logs the details of a vote.
+func (cs *State) logVoteDetails(vote *types.Vote) {
```
Comment (cason) [inline internal/consensus/state.go:2304]: I don't see the point of moving this to a method.

File: internal/consensus/state.go [hunk 3]
```diff
@@ -2264,54 +2311,59 @@ func (cs *State) addVote(vote *types.Vote, peerID p2p.ID) (added bool, err error
 		"extLen", len(vote.Extension),
 		"extSigLen", len(vote.ExtensionSignature),
 	)
+}
 
-	if vote.Height < cs.Height || (vote.Height == cs.Height && vote.Round < cs.Round) {
-		cs.metrics.MarkLateVote(vote.Type)
-	}
+// isLateVote checks if a vote is a late vote.
+func (cs *State) isLateVote(vote *types.Vote) bool {
```
Comment (cason) [inline internal/consensus/state.go:2317]: This is harder to read than the original version.

The same for the similar helpers. I think that the content of the ifs should be maintained for better readability. A comment can do the work of explaining what is being evaluated.

File: internal/consensus/state.go [hunk 4]
```diff
@@ -2264,54 +2311,59 @@ func (cs *State) addVote(vote *types.Vote, peerID p2p.ID) (added bool, err error
...
-		added, err = cs.LastCommit.AddVote(vote)
-		if !added {
-			// If the vote wasn't added but there's no error, its a duplicate vote
-			if err == nil {
-				cs.metrics.DuplicateVote.Add(1)
-			}
-			return added, err
-		}
+// handlePrecommitForPreviousHeight handles a precommit vote for the previous height.
+func (cs *State) handlePrecommitForPreviousHeight(vote *types.Vote) (bool, error) {
```
Comment (cason) [inline internal/consensus/state.go:2327]: This could be something like handleLastCommitVote.

File: internal/consensus/state.go [hunk 5]
```diff
@@ -2264,54 +2311,59 @@ func (cs *State) addVote(vote *types.Vote, peerID p2p.ID) (added bool, err error
...
+	if cs.Step != cstypes.RoundStepNewHeight {
```
Comment (cason) [inline internal/consensus/state.go:2328]: I know this was in the original code, but I don't see much point in this test with debug-only purposes.

We only add votes to `cs.LastCommit` when it is a precommit from the previous height **and** we are in the new_height round step.

File: internal/consensus/state.go [hunk 6]
```diff
@@ -2254,7 +2254,54 @@ func (cs *State) tryAddVote(vote *types.Vote, peerID p2p.ID) (bool, error) {
...
+
+	// A precommit for the previous height?
+	// These come in while we wait timeoutCommit
+	if cs.isPrecommitForPreviousHeight(vote) {
+		return cs.handlePrecommitForPreviousHeight(vote)
+	}
+
+	// Height mismatch is ignored.
+	// Not necessarily a bad peer, but not favorable behavior.
+	if cs.isHeightMismatch(vote) {
```
Comment (cason) [inline internal/consensus/state.go:2273]: How is this code easier to read than `vote.Height != cs.Height`?

File: internal/consensus/state.go [hunk 7]
```diff
@@ -2264,54 +2311,59 @@ func (cs *State) addVote(vote *types.Vote, peerID p2p.ID) (added bool, err error
...
+		cs.Logger.Debug("precommit vote came in after commit timeout and has been ignored", "vote", vote)
+		return false, nil
+	}
 
-		cs.Logger.Debug("added vote to last precommits", "last_commit", cs.LastCommit.StringShort())
-		if err := cs.eventBus.PublishEventVote(types.EventDataVote{Vote: vote}); err != nil {
-			return added, err
+	added, err := cs.LastCommit.AddVote(vote)
+	if !added {
+		// If the vote wasn't added but there's no error, it's a duplicate vote
```
Comment (cason) [inline internal/consensus/state.go:2336]: There is a `handleDuplicateVote` method...

**Review by cason — COMMENTED** [review]
Thank you for this.

While I agree that some refactoring would increase readability, I would prefer moving outside from `addVote` only auxiliary logic.

**Comment by faddat** [comment]
@cason - appreciate your comments, and will give this a proper go-through.

File: internal/consensus/state.go (hunk 1 above)
Comment (faddat) [reply internal/consensus/state.go:2423]: will check that.

File: internal/consensus/state.go (hunk 2 above)
Comment (faddat) [reply internal/consensus/state.go:2302]: Moving this out makes a lot of sense to me.  It takes something that isn't "core" and moves it out.  What am I missing?

File: internal/consensus/state.go (hunk 1 above)
Comment (faddat) [reply internal/consensus/state.go:2423]: @melekes it's safe to say, that if we are ignoring that error, it should be found by a test, right?

Would it make sense to address this by adding a test?

File: internal/consensus/state.go (hunk 6 above)
Comment (faddat) [reply internal/consensus/state.go:2271]: english.  This version calls a method whose name is self descriptive.  

But to change it back is nbd

**Comment by faddat** [comment]
Proposal: we make addvote its own file

File: internal/consensus/state.go (hunk 4 above)
Comment (faddat) [reply internal/consensus/state.go:2325]: gotcha, makes sense.

File: internal/consensus/state.go [hunk 8]
```diff
@@ -2346,124 +2398,122 @@ func (cs *State) addVote(vote *types.Vote, peerID p2p.ID) (added bool, err error
...
+			cs.enterPrevote(cs.Height, cs.Round)
+		}
 	}
+}
 
-	return added, err
+// handlePrecommit handles a precommit.
+func (cs *State) handlePrecommit(vote *types.Vote) {
+	precommits := cs.Votes.Precommits(vote.Round)
+	cs.Logger.Debug("added vote to precommit",
```
Comment (faddat) [inline internal/consensus/state.go:2492]: this here could also be its own method

File: internal/consensus/state.go (hunk 1 above)
Comment (faddat) [reply internal/consensus/state.go:2423]: we definitely are.  Nice catch!  Fixing...

File: internal/consensus/state.go (hunk 2 above)
Comment (faddat) [reply internal/consensus/state.go:2304]: unresolve if you really want this one changed.  Some of the others were very good catches, thank you!

**Comment by ValarDragon** [comment]
I recommend we close this and do some simpler PR's to reduce complexity. It may be helpful to examine some of the simplification as what can be re-used between the reactor and state.go

**Comment by melekes** [comment]
> I recommend we close this and do some simpler PR's to reduce complexity. It may be helpful to examine some of the simplification as what can be re-used between the reactor and state.go

Agree. @faddat please let us know if you want to break this work into smaller PRs as per @ValarDragon comment.

**Pull request closed without merging**